core/vm/lightclient/ics23_proof.go

```diff
@@ -0,0 +1,210 @@
+	}
+	// Only support an existence proof or nonexistence proof (batch proofs currently unsupported)
+	switch len(args) {
+	case 0:
+		// Args are nil, so we verify the absence of the key.
+		absent := ics23.VerifyNonMembership(op.Spec, root, op.Proof, op.Key)
+		if !absent {
+			return nil, fmt.Errorf("proof did not verify absence of key: %s", string(op.Key))
+		}
+
+	case 1:
+		// Args is length 1, verify existence of key with value args[0]
+		if !ics23.VerifyMembership(op.Spec, root, op.Proof, op.Key, args[0]) {
+			return nil, fmt.Errorf("proof did not verify existence of key %s with given value %x", op.Key, args[0])
+		}
+	default:
+		return nil, fmt.Errorf("args must be length 0 or 1, got: %d", len(args))
+	}
+
+	return [][]byte{root}, nil
+}
+
+// ProofOp implements ProofOperator interface and converts a CommitmentOp
+// into a merkle.ProofOp format that can later be decoded by CommitmentOpDecoder
+// back into a CommitmentOp for proof verification
+func (op CommitmentOp) ProofOp() merkle.ProofOp {
+	bz, err := op.Proof.Marshal()
+	if err != nil {
+		panic(err.Error())
+	}
+	return merkle.ProofOp{
+		Type: op.Type,
+		Key:  op.Key,
+		Data: bz,
+	}
+}
+
+func Ics23ProofRuntime() (prt *merkle.ProofRuntime) {
+	prt = merkle.NewProofRuntime()
+	prt.RegisterOpDecoder(ProofOpIAVLCommitment, CommitmentOpDecoder)
+	prt.RegisterOpDecoder(ProofOpSimpleMerkleCommitment, CommitmentOpDecoder)
+	return
+}
+
+func VerifyValue(root []byte, version int64, proof *merkle.Proof, keyPath string, value []byte) error {
+	poz, err := Ics23ProofRuntime().DecodeProof(proof)
+	if err != nil {
+		return errors.Wrap(err, "decoding proof")
+	}
+
+	if len(poz) != 2 {
+		return fmt.Errorf("length of proof ops should be 2")
+	}
+
+	keys, err := merkle.KeyPathToKeys(keyPath)
+	if err != nil {
+		return fmt.Errorf("get keys error, err=%s", err.Error())
+	}
+
+	if len(keys) != 2 {
+		return fmt.Errorf("length of keys should be 2")
+	}
+
+	storeKey, valueKey := keys[0], keys[1]
+
+	iavlPo := poz[0]
+	if iavlPo.ProofOp().Type != ProofOpIAVLCommitment {
+		return fmt.Errorf("invalid proof op type, should be  %s", ProofOpIAVLCommitment)
+	}
+
+	if !bytes.Equal(valueKey, iavlPo.GetKey()) {
+		return fmt.Errorf("invalid proof of key, require %X, got %X", iavlPo.GetKey(), valueKey)
+	}
+
+	iavlRoot, err := iavlPo.Run([][]byte{value})
+	if err != nil {
+		return fmt.Errorf("invalid iavl proof, err=%s", err.Error())
+	}
+
+	if len(iavlRoot) != 1 {
+		return fmt.Errorf("invalid return of iavl proof")
+	}
+
+	// calculate store hash
+	storeInfo := StoreInfo{
+		Name: string(storeKey),
+		Core: StoreCore{
+			CommitID: CommitID{
+				Version: version,
+				Hash:    iavlRoot[0],
+			},
+		},
+	}
+
+	storeHash := storeInfo.Hash()
+
+	simplePo := poz[1]
+	if !bytes.Equal(storeKey, simplePo.GetKey()) {
+		return fmt.Errorf("invalid proof of key, require %X, got %X", simplePo.GetKey(), storeKey)
+	}
+	storeRoot, err := simplePo.Run([][]byte{storeHash})
```

Comment:
Can we assert the type of `simplePo` to be `SimpleValueOp`?   Two `iavlOp`  is not allowed here,